### ISSUE TITLE
Incorrect gain due to info["nTelegraphEnable"] in ABF 1.3 files

### ISSUE THREAD

**Issue by sigmareceptor** [NeuralEnsemble/python-neo#544]
Hi all,

This is my first time posting on Github, so excuse me if I break any conventions. Anyway, I've been trying to read in .ABF files using both AxonIO and AxonRawIO. My ABF files are created in a somewhat wonky way. I record using a HEKA EPC10 amp, then convert the files from extracted .itx files to .ABF, so I can use Clampfit. To convert them, I use the [ABF utility from Synaptosoft](http://www.synaptosoft.com/MiniAnalysis/ABFUtility.html). However, this piece of software is old as heck, so it converts my .itx files to .ABF version 1.3. 

I've noticed that when I read in my files using either AxonIO or AxonRawIO, the values are way off. The files themselves are integer-based, and in the conversion of integer to floats, something goes wrong. I looked at the code, and it seems that the problem has to do with the gain calculation. Specifically, if I check the info as used by AxonRawIO to calculate the gain (see lines 176 to 186 of neo/rawio/axonrawio.py), the program checks whether info['nTelegraphEnable'] exists. The odd thing is that the info file does contain this variable (specifically, printing info['nTelegraphEnable'] yields an array like the following: array([-7220, -7168, -7137, -7138, -7173, -7231, -7298, -7374, -7454, -7522, -7565, -7577, -7571, -7549, -7520, -7486]). When I open the .ABF file with Clampfit and check the metadata, it explicitly says "Telegraphs disabled". Moreover, when I manually calculate the gain, and forego the division by info['fTelegraphAdditGain'], I get the correct gain, and my files work correctly. 

Anyway, I presume this is a bug? Or is .ABF v1.3 too old to be supported by python-neo?

An example .ABF file can be found [here](https://www.dropbox.com/s/8g5x2zd5qb3yuhi/130618%201-12.ABF?dl=0). Please note that this is not the exact file I used to generate the array mentioned above. However, the problems are the exact same, as is the solution.

Cheers.

**Comment by samuelgarcia** [comment]
I guess it is a bug.
I will have look.
Normally neo.AxonIo support both ABF1 and ABF2.

**Comment by samuelgarcia** [comment]
I confirm, this is strange the nTelegraphEnable vector should be 1 or 0 and it is some random numbers.

2 hypothesis:
  1. header in ABF 1.3 is different from the spec we have coded (see headerDescriptionV1)
  2. "ABF utility from Synaptosoft" have also a bug when writing the field nTelegraphEnable.

Hypothesis 2 is not very credible because in that camplfit should also fail.

**Comment by samuelgarcia** [comment]
@swharden : here a new challenge, that we can solve both in pyabf and neo.
Here a file version 1.3 come from a converting tools.
It can be open with clmapfit without problem.
In neo the magnitude is wrong due to 'nTelegraphEnable' which is supposed to be 1 or 0 in our implementation. I guess it is the same in pyabf.
Do you have any idea.

**Comment by swharden** [comment]
When I open the OP's file in ClampFit, this first values are around -200 pA. When I open this file with pyABF the first few values are around -20 pA. Since both values seem reasonable, ***can somebody verify which values are actually the correct values from the original ITX files?***

I agree parts of the header from the ABF created by the [SynaptoSoft ABF Utility](http://www.synaptosoft.com/MiniAnalysis/ABFUtility.html) are buggy (including nTelegraphEnable), but I don't want to get sucked into a rabbit hole of figuring out what the problem is if Neo and pyABF are actually reporting the correct values and a buggy header is causing ClampFit to report the wrong ones.

ClampFit values | pyABF values
--- | ---
![image](https://user-images.githubusercontent.com/4165489/42413888-e0b3739e-81f7-11e8-8bea-5bd7ada20811.png) | ![image](https://user-images.githubusercontent.com/4165489/42413938-34237974-81f9-11e8-8761-7694cca87197.png)

pyABF code for reference
```python
import pyabf
abf = pyabf.ABF("130618-1-12.abf")
plt.plot(abf.sweepX, abf.sweepY)
plt.title(abf.abfID)
plt.ylabel(abf.sweepLabelY)
plt.xlabel(abf.sweepLabelX)
plt.show()
```

**Comment by swharden** [comment]
Actually, I think there is a simple solution here. `nTelegraphEnable` should only contain 0s and 1s. Anything else is an improperly-formatted header. A simple fix to support all cases is to specifically look for a 1 (which indicates we need to further scale the data by `fTelegraphAdditGain`). Anything else (usually 0, or a crazy number in this case) just ignores `fTelegraphAdditGain`.

Since this problem is an invalid ABF1 header caused by the SynaptoSoft ABF Utility, one wouldn't be faulted for not doing anything to support ABF files with invalid headers. However, here's an easy fix:

Modify [axonrawio.py#L182-L183](https://github.com/NeuralEnsemble/python-neo/blob/master/neo/rawio/axonrawio.py#L182-L183):

```python
if info['nTelegraphEnable'][chan_id]:
	gain /= info['fTelegraphAdditGain'][chan_id]
```

To look like this:

```python
if info['nTelegraphEnable'][chan_id]!=0:
	gain /= info['fTelegraphAdditGain'][chan_id]
```

It worked for me with pyABF:

![figure_2](https://user-images.githubusercontent.com/4165489/42414243-236f3342-81ff-11e8-8d4b-37eb8a5365a2.png)

**Comment by samuelgarcia** [comment]
Dear @swharden.
Thank you very much for your expertise in abf format and participation in solving bugs in neo.
The patch looks resonnable.
Maybe we could add a warning when this nTelegraphEnable is buggy like this:

```
if not np.all(np.in1d(info['nTelegraphEnable'], [0,1]):
    logger.warning('The ABF1 header is probably buggy, nTelegraphEnable have strange values')
```

**Comment by swharden** [comment]
Hi @samuelgarcia, 

I modified the pull to add a warning based on your suggestion. It only checks for invalidly-formatted headers in ABF version 1 files, but this seems fine since those are likely the only type to be synthesized from scratch using third party utilities.

I liked your suggestion of using `np.all` and `np.in1d`, but I find this to flow a bit more logically and to be easier to read. What do you think of the change? 

https://github.com/NeuralEnsemble/python-neo/pull/547/files

```python
if info['nTelegraphEnable'][chan_id] == 0:
    pass
elif info['nTelegraphEnable'][chan_id] == 1:
    gain /= info['fTelegraphAdditGain'][chan_id]
else:
    logger.warning('ignoring buggy nTelegraphEnable')
```

**Comment by samuelgarcia** [comment]
closed by #547
